### ISSUE TITLE
Investigate if signal names may be reused in DBC files

### ISSUE THREAD

**Issue by erikbosch** [eclipse/kuksa.val.feeders#122]
A CAN databse consist of a number of messages, each containing a number of signals. At the moment the DBC Feeder design is based on the assumption that signal names are unique. We should better investigate if that is true, or be prepared that it maybe is not always true. According to https://support.vector.com/kb?id=kb_article_view&sysparm_article=KB0011320&sys_kb_id=4ae40ea41b2614148e9a535c2e4bcba7&spa=1 it seems names may be reused in different frames.

A few possible solutions:

**Alt 1: Treat it as a limitation**

State that behavior if there are multiple signals with the same name is undefined. We could refactor `get_canid_for_signal` so that it the first time it search for a signal name does not stop when finding first match, but rather continue and give a warning or error if it finds a second instance. So if you do not get a warning or error you are fine.

**Alt 2: Mandatory message name**

Make message name mandatory in the mapping, make sure it matches.

**Alt 3: Optional message name**

Only require message name if there are conflicts. I.e. if there are 2+ signals with same name give an error if no message name has been specified. if a message name has been specified verify that it match

**Comment by erikbosch** [comment]
FYI @sophokles73 @SebastianSchildt  I would prefer Alt3

**Comment by sophokles73** [comment]
I have conducted some tests with a database file that contains two messages defining a signal of the same name.
The dbparser correctly loads the database and you can look up both messages by name which then both contain a signal of then given name.

I agree that Alt 3 is the least intrusive approach. However, based on the assumption that a signal is only correctly identified by the message name + signal name, I would propose to define this as the default way of specifying a signal and to consistently use this way in all examples. We could then mention that for backwards compatibility reasons it is possible to omit the message name if the signal name is unique across the database.

**Comment by SebastianSchildt** [comment]
**Alt 3** is my vote as well.

As for modifying the example mapping I am on the fence, as I guess due to laziness/usability and the fact most people will map from one small(ish) DBCs, probably the most common, easy pattern is just using signal names.

So what I am saying is: Definitely describe the implications of only using signal name, but not complain/threat too much wrt to warnings during runtime. Perfect would be, if on startup after loading dbc and mapping we can make a check across the database only for uniqueness of signals used in mapping, and if we find an ambiguous case terminate or warn that behaviour is undefined.

**Comment by sophokles73** [comment]
I see (at least) the following use cases:
1. Use a generic DBC file that contains multiple (identical) message definitions for different frame IDs.
   
   An example of this would be to use a DBC file that contains j1939 message definitions with frame IDs specific to multiple OEMs. Such a file can be used to *read* signals from the CAN bus and map them uniquely to corresponding VSS signals. However, mapping VSS signals to CAN messages would either require unique signal names to be used in the DBC file or to also specify the message name in the *vss2dbc* mapping information.
2. Use a DBC file that is specific to a particular OEM/vehicle and contains only unique message definitions with frame IDs specific to the given OEM/vehicle.
   
   In this case bidirectional mapping between CAN frames and VSS signals should be possible, assuming that signal names are unique across the message definitions in the DBC file.

Based on the *--dbc2val* and *--val2dbc* command line switches we could assert the requirements during startup and support both cases accordingly. Given that the second use case only really makes sense in an OEM/vehicle specific scenario anyway, I would suggest to simply fail starting the feeder if the DBC file contains duplicate signal names.

WDYT? @SebastianSchildt @erikbosch

**Comment by sophokles73** [comment]
@SebastianSchildt @erikbosch any thoughts?

**Comment by SebastianSchildt** [comment]
I am not sure I follow the lined out action completely. Let me rephrase what I got, and what makes sense to me:

 - technically speaking a message/Signal id in DBC is not necessarily  unique without the frame id, although for most practical purposes it is
 - if there are duplicate signal ids and on of those is used in the mapping we could find out during startup

That means for "val2dbc" mappings:
 - Iff  there is a mapping that refers to one doubled signal, you would not know the frame id to create, so we should fail in that case. (If there are "just" double signals in DBC we do not use, we do not need to care)

For dbc2val mappings:
 - Shouldn't we also fail? Or at least give a strong warning? Sure, both would be mapped, but it might point to a user error, because you can not be sure, the two signals are really semantically exactly the same, i.e. configuring it that way may point out a user error in the mapping file. Assume two things called "VSpeed", where one is already averaged over 2 seconds and another one is raw. That would look weird for apps when the provider is mapping those simultaneously to the same VSS point.


If that is roughly enough, would you agree for a (first) PR that is sufficient, with the disadvantage though that to "repair" such an occurence user side, one would need to modify the DBC.

Or would you suggest we should support optionally specifying frame ids in both mapping directions in addition to signal? I think supporting this would be some "legwork" that needs to be done in code, and I am not currently sure how common the situations you lined out in the examples are. It would however be the most foolproof solution.

Did I parse the situation ad proposed solutions correctly @sophokles73 ?

Generally I would say, we'd gladly accept PRs implementing parts/all of this. 

Otherwise, as a minimum I think we should at least implement a check whether overloaded signals exist in dbc file (and are used in mapping) and print a strong warning, to help debugging in such corner cases. So we should treat this issue as "bug" until we have that

**Comment by sophokles73** [comment]
> Otherwise, as a minimum I think we should at least implement a check whether overloaded signals exist in dbc file (and are used in mapping) and print a strong warning, to help debugging in such corner cases. 

FMPOV that would be reasonable. As pointed out before, I also believe that we should actually exit with an error in such a case if the *--val2dbc* switch has been set.

**Comment by erikbosch** [comment]
We have migrated CAN Provider (dbcfeeder) to a new [repo](https://github.com/eclipse-kuksa/kuksa-can-provider). Long term my intention is to close all dbc/can-related issues in this repository. If we see the issue as still interesting or relevant we could create a new issue in the new repo, possibly referencing the old issue in the repository.

Suggested way forward:

- I will migrate selected issues early 2024
- If this issue has not been migrated but you want it to remain open please comment.
- Remaining issues no one seems to be interested in will be closed around 2024-03-31

**Comment by sophokles73** [comment]
@erikbosch I also believe that it would be beneficial to continue work on this in the new repo. Can you please migrate it?

**Comment by erikbosch** [comment]
I copied this issue to https://github.com/eclipse-kuksa/kuksa-can-provider/issues/8. Old comments remain here though. Closing this one. (FYI @sophokles73 )
